refactor(parser): remove lookahead usage in parsing arrow function expressions

camchenry opened the pull request:
- towards https://github.com/oxc-project/oxc/issues/11194

Replaces all usage of the lexer lookahead functions like `nth_kind`, so that we only use checkpoints and rewinding, which is aligned with the TypeScript compiler implementation. This also will allow us to remove the lookahead field from the lexer in the future.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11220?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11220** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11220?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11220?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11218** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11218?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11220).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry commented:
@Boshen is this sort of snapshot update expected? It looks like the parser error message has changed here, where before we just had "unexpected token" and now it says we were looking for a "=>", which feels better? But I'm not sure if it's correct or not, since I'm less familiar with the parser changes.


<img width="1262" alt="image" src="https://github.com/user-attachments/assets/eef76499-3dc0-4112-bbc1-558fc985a0d7" />

camchenry commented:
Actually, it looks like this is correct I think. This is the same error that TypeScript will emit:

<img width="384" alt="image" src="https://github.com/user-attachments/assets/e108eadd-922e-4d4a-b6a9-27f1fbad294a" />

https://www.typescriptlang.org/play/?#code/BQOjEMEoAIGpoIwCgg

Boshen commented:
> @Boshen is this sort of snapshot update expected? It looks like the parser error message has changed here, where before we just had "unexpected token" and now it says we were looking for a "=>", which feels better? But I'm not sure if it's correct or not, since I'm less familiar with the parser changes.
> 
> <img alt="image" width="1262" src="https://private-user-images.githubusercontent.com/1514176/446166532-eef76499-3dc0-4112-bbc1-558fc985a0d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4ODI4ODksIm5iZiI6MTc0Nzg4MjU4OSwicGF0aCI6Ii8xNTE0MTc2LzQ0NjE2NjUzMi1lZWY3NjQ5OS0zZGMwLTQxMTItYmJjMS01NThmYzk4NWEwZDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MjJUMDI1NjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWJjZGRiOGFhOGMwNGMyZjVlMDdhMmRhNzUwMjE3NWE1MTk0MTVkYWI1YWIwMTFjMzJhNDUyYTJmYThhMDAwNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.r9e6LfeaEXQDSc4ALFu1ugqGi9-DMAQHUJ4h5GCF04g">

Yes! We expect more alignment to tsc from making these changes.

Boshen commented:
### Merge activity

* **May 22, 2:58 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(parser): remove lookahead usage in parsing arrow function expressions (#11220)

- towards https://github.com/oxc-project/oxc/issues/11194

Replaces all usage of the lexer lookahead functions like `nth_kind`, so that we only use checkpoints and rewinding, which is aligned with the TypeScript compiler implementation. This also will allow us to remove the lookahead field from the lexer in the future.
